### ISSUE TITLE
[aws-rds] What's the best way to define an Aurora DbCluster so that is can survive destroy/re-create cycles?

### ISSUE THREAD

**Issue by mcalello** [aws/aws-cdk#9304]
## :question: General Issue
I'm using aurora-postgresql serverless 10.7, so we are currently using L1 constructs to define our CfnDBCluster.  What is the best way to arrange our CDK code to perform the following:

- Create this DbCluster from scratch when brand new with its own generated Secret (we have this part working already)
- Snapshot the Data prior to deleting it on Destroy.
- Use the previously saved snapshot to repopulate the data on recreate.
- Rewire a Secret to the recreated DbCluster.

### The Question
We've encountered a couple of snags attempting to automate the resurrection of our Aurora Postgres DbCluster through the destroy/re-create cycle.  I'll enumerate some of my confusion in trying to get this to work, but ultimately the main ask here if for the Best Practice for handling this case.

- If CfnDBCluster's snapshot_identifier requires an existing snapshot (DBClusterSnapshotNotFoundFault), should we deal with the base case of creating the DbCluster from scratch first? If, so how?
-  CfnDBCluster throws an error when both snapshot_identifier and master_username are provided, "Exactly one of SnapshotIdentifier and MasterUsername must be specified".  So when a snapshot is provided, what is the best way to re-establish it's Secret?
- Our generated secret gets destroyed along with the Destroy.  Should we change this to Retain?
- At least one 3rd party CDK blog article from earlier this year references a CfnDBCluster attribute called finalSnapshotIdentifier. Was this an error, or was a finalSnapshotIdentifier removed from CfnDBCluster?
- Is there a way to specify the name for the snapshot taken prior to destroy for a DbCluster?

**Comment by skinny85** [comment]
Hello @mcalello ,

thanks for opening the issue. 

First, let me clarify a bit. From what I can tell, none of your questions are CDK-specific (unless I'm missing something). Are questions in general about RDS, and cluster snapshots?

**Comment by mcalello** [comment]
Sure - we could be asking the same questions of CloudFormation, or RDS specifically - but without the potential of the higher-level patterns available in CDK. But given a desire to embrace CDK for our IaC, I'm asking for the best approach to define our DbCluster to tolerate an occasional destroy/recreate cycle.

**Comment by srethira** [comment]
@skinny85 This is kind of related to [this](https://github.com/aws/aws-cdk/issues/4379) feature. Until we have a higher level construct like DbClusterFromSnapshot. What is the recommended way of doing it using the L1 construct? Is there any example?

**Comment by susguzman** [comment]
While the option to restore from a cluster is available, we can use L1 construct (CloudFormation resources). Here is an example using python:

```python
from aws_cdk import (
    aws_ec2 as ec2,  # pip install aws_cdk.aws_ec2
    aws_rds as rds,  # pip install aws_cdk.aws_rds
    core
)  # pip freeze > requirements.txt


class RDSStack(core.Stack):

    def __init__(self, scope: core.Construct, id: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        vpc = ec2.Vpc(self, "VPC")

        rds_subnet_group = rds.CfnDBSubnetGroup(self, 'rdsSubnetGroup',
                                                db_subnet_group_description='private subnets for rds',
                                                subnet_ids=vpc.select_subnets(
                                                    subnet_type=ec2.SubnetType.PRIVATE).subnet_ids)

        aurora_cluster = rds.CfnDBCluster(self, 'auroraCluster',
                                          engine='aurora-mysql',
                                          engine_version='5.7.mysql_aurora.2.08.1',
                                          db_cluster_parameter_group_name='default.aurora-mysql5.7',
                                          snapshot_identifier='<snapshot_arn>', #your snapshot
                                          db_subnet_group_name=core.Fn.ref('rdsSubnetGroup'))

        aurora_instance = rds.CfnDBInstance(self, 'aurora-instance',
                                            db_cluster_identifier=core.Fn.ref('auroraCluster'),
                                            engine='aurora-mysql',
                                            engine_version='5.7.mysql_aurora.2.08.1',
                                            db_instance_class='db.t3.small',
                                            db_parameter_group_name='default.aurora-mysql5.7')
```

**Comment by mcalello** [comment]
Thanks @susguzman for the example.  I was able to get something that could run both from scratch, and then during re-create - but I needed to rely on boto3/aws-sdk calls during the synth phase to determine which direction to go.  I had to look for a retained previous Secret by friendly name, and also an existing manual snapshot to decide if it was a recreate or create from scratch.

I'm still curious - what are others doing to maintain an Aurora cluster for production and handle the cases where the cluster needs to be destroyed?

**Comment by followben** [comment]
@mcalello - I landed here with the same question. Feels odd there's no simple answer for an in-place restore via higher-level constructs in the CDK. Though #10130 has landed, seems we still need to use the SDK to manually query and decide what to do (we're still reliant on L1 constructs anyway as we're Aurora Serverless).

Still, at least I've a way forward based on your comments and @susguzman's example. If you have any code you'd like to share, I'd be very grateful!

**Comment by agalazis** [comment]
even with #10130 how about credentials and storageEncrypted options? can they be reset?

**Comment by agalazis** [comment]
ok I see that credentials is a thing:
https://github.com/aws/aws-cdk/blob/eb45ca816626b243daacbd3a8916ac1e5db202ea/packages/%40aws-cdk/aws-rds/test/test.instance.ts#L322

**Comment by skinny85** [comment]
@agalazis Importantly, you can't change the master username of the database when creating it from a snapshot (the [`SnapshotCredentials` class reflects this](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-rds.SnapshotCredentials.html)).

**Comment by agalazis** [comment]
The above is only an option for instance from snapshot not cluster

**Comment by jacobevans** [comment]
@mcalello Did you ever figure out a good pattern for doing this?  It's now pretty easy to build a new cluster from a snapshot with `DatabaseClusterFromSnapshot`, but I still don't see a good option for the bits about keeping track of the final snapshot from the previous instance, and using that one (or starting from scratch if it doesn't exist).

**Comment by followben** [comment]
@jacobevans We already a wrapper script that invoked the CDK as part of our CI pipeline, so I modified that to look for an environment variable with a snapshot ARN (which got updated at the end of every deploy via the SDK). If it found one, it used that as the basis for the cluster instead of starting greasy & post deploy did the work of updating the master user credentials (again, using the SDK). Was fiddly to write, but suited our needs. Also meant we could specify which snapshot to use if we wanted to revert to a different state.

Silly thing is I've now changed jobs and my new company needs exactly the same functionality, albeit in Python rather than JS! 

Thinking about how the services work under the hood, I understand why the CDK (and indeed RDS) don't have a simple answer here. But as a platform user it feels like a real omission.
